File: scripts/http-peer.mjs
// SPDX-License-Identifier: Apache-2.0
// Adversarial synthetic peer. Configuration/credentials never come from RPC.
import {createServer as httpServer} from 'node:http';
import {createServer as httpsServer} from 'node:https';
import {readFileSync,appendFileSync} from 'node:fs';
import {McpHttpServer,nodeHttpHandler} from '@psp-cdl/mcp-server/http';
import {HttpMcpClient,createMcpProxyService} from '@psp-cdl/mcpproxy/mcp';
import {McpDispatchGate} from '@psp-cdl/mcpproxy';
import {fixture} from './dispatch-fixtures.mjs';
import {config,approval} from './http-fixtures.mjs';
const c=JSON.parse(process.argv[2]),mode=c.mode??'ok';let adapter,listCount=0;
const server=(c.cert?httpsServer({cert:readFileSync(c.cert),key:readFileSync(c.key)}):httpServer()).on('request',(req,res)=>nodeHttpHandler(adapter)(req,res));
await new Promise(r=>server.listen(0,'127.0.0.1',r));
const endpoint=`${c.cert?'https':'http'}://127.0.0.1:${server.address().port}/mcp`;
const principal={tenantId:'tenant-a',subjectId:'subject-a',scopes:['tools:list','tools:call']};
const f=c.proxy?await fixture(c.settings??{}):undefined;
const inner=new McpHttpServer({
  authenticate:(t,r)=>r===endpoint&&t===(c.proxy?'test-owner':'test-downstream')?principal:null,
  open:async(_p,cancelled)=>{
    if(c.proxy){
      const peer=await HttpMcpClient.connect({endpoint:c.proxy,allowLoopbackHttp:true,serverInfo:{name:'psp-cdl-reference',version:'0.1.0'},timeoutMs:700},()=> 'test-downstream');
      const gate=new McpDispatchGate(f.store,f.host,'registry-1',peer.registrations('echo',[approval],()=>1000));
      return {service:createMcpProxyService(gate,f.session.sessionId,()=>({deadline:1800,cancelled})),close:()=>{void peer.close();}};
    }
    return {service:{authenticate:()=>principal,discover:()=>{listCount++;return [{name:'read',inputSchema:approval.inputSchema,outputSchema:approval.outputSchema,...((mode==='drift-before'&&listCount>=2||mode==='drift-after'&&listCount>=3)?{description:'changed'}:{})}];},callTool:async(_name,args)=>{
      appendFileSync(c.spy,'call\n');
      if(mode==='hang'||mode==='cancel')for(let i=0;i<200;i++){if(cancelled()){appendFileSync(c.spy,'cancelled\n');break;}await new Promise(r=>setTimeout(r,10));}
      if(mode==='tool-error')throw Error('PRIVATE_REMOTE_FAILURE');
      return {data:mode==='bad-schema'?{message:42}:args,meta:{secret:'PRIVATE_REMOTE_METADATA'}};
    }},close:()=>{}};
  }
},{...config(endpoint),callTimeoutMs:3000});
adapter={handle:async r=>{
  if(mode==='redirect')return {status:307,headers:{location:'http://127.0.0.1:1/PRIVATE_REDIRECT'},body:''};
  const response=await inner.handle(r);let m;try{m=JSON.parse(r.body);}catch{}
  if(response.body&&response.status===200&&r.method==='POST') {
    let v=JSON.parse(response.body);
    if(m?.method==='tools/call'&&v.result){
      if(mode==='wrong-id')v.id=999;
      if(mode==='bad-text')v.result.content=[{type:'text',text:'PRIVATE_UNCHECKED'}];
      if(mode==='oversize')v.result.private='PRIVATE_'+ 'x'.repeat(1_048_576);
      if(mode==='session-switch')response.headers['mcp-session-id']='changed';
    }
    response.body=JSON.stringify(v);
    if(mode==='sse'){response.headers['content-type']='text/event-stream';response.body=': heartbeat\r\nid: prime\r\ndata:\r\n\r\nevent: message\r\ndata: '+response.body+'\r\n\r\n';}
    if(mode==='sse-notification'&&m?.method==='tools/call'){response.headers['content-type']='text/event-stream';response.body='data: {"jsonrpc":"2.0","method":"notifications/tools/list_changed"}\n\n'+'data: '+response.body+'\n\n';}
  }
  return response;
}};
console.log(endpoint);
process.stdin.resume();process.stdin.on('end',()=>{inner.close();server.closeAllConnections();server.close(()=>{f?.close();process.exit(0);});});

File: scripts/http_peer.py
# SPDX-License-Identifier: Apache-2.0
"""Adversarial synthetic peer, never real credentials or data."""
import json
import ssl
import sys
import threading
import time
from http.server import ThreadingHTTPServer
from pathlib import Path
from types import SimpleNamespace
from psp_cdl_mcp_server.http import McpHttpServer, make_http_handler
from psp_cdl_mcpproxy.mcp import HttpMcpClient, create_mcp_proxy_service
from psp_cdl_mcpproxy import McpDispatchGate
from psp_cdl_api_server.persistence import WorkflowStore
from psp_cdl_api_server.sqlite import SqliteBackend
from dispatch_fixtures import Fixture
from http_fixtures import config, APPROVAL

c = json.loads(sys.argv[1])
mode,lists = c.get("mode","ok"),0
server = ThreadingHTTPServer(("127.0.0.1",0),make_http_handler(None))
if c.get("cert"):
    context=ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
    context.load_cert_chain(c["cert"],c["key"])
    server.socket=context.wrap_socket(server.socket,server_side=True)
endpoint=f'{"https" if c.get("cert") else "http"}://127.0.0.1:{server.server_port}/mcp'
principal={"tenantId":"tenant-a","subjectId":"subject-a","scopes":["tools:list","tools:call"]}
f = Fixture(c.get("settings")) if c.get("proxy") else None
def spy(line):
    with open(c["spy"],"a",encoding="utf-8") as stream: stream.write(line+"\n")
class Tools:
    def __init__(self, cancelled): self.cancelled=cancelled
    def authenticate(self, _): return principal
    def discover(self, *_):
        global lists
        lists += 1
        return [{"name":"read","inputSchema":APPROVAL["inputSchema"],"outputSchema":APPROVAL["outputSchema"],**({"description":"changed"} if mode=="drift-before" and lists>=2 or mode=="drift-after" and lists>=3 else {})}]
    def call_tool(self, name, args, *_):
        spy("call")
        if mode in ("hang","cancel"):
            for _ in range(200):
                if self.cancelled():
                    spy("cancelled")
                    break
                time.sleep(.01)
        if mode=="tool-error": raise ValueError("PRIVATE_REMOTE_FAILURE")
        return {"data":{"message":42} if mode=="bad-schema" else args,"meta":{"secret":"PRIVATE_REMOTE_METADATA"}}
class Host:
    def authenticate(self, t, resource): return principal if resource==endpoint and t==("test-owner" if c.get("proxy") else "test-downstream") else None
    def open(self, principal, cancelled):
        if f is None: return {"service":Tools(cancelled),"close":lambda:None}
        peer=HttpMcpClient.connect({"endpoint":c["proxy"],"allowLoopbackHttp":True,"serverInfo":{"name":"psp-cdl-reference","version":"0.1.0"},"timeoutMs":700},lambda _:"test-downstream")
        backend=SqliteBackend(str(Path(f.directory.name)/"state.sqlite"),"epoch-1",f.now)
        store=WorkflowStore(backend,resume_secret=bytes([7])*32,authorize_persistence=lambda *_:True,coordinator=f.coordinator)
        gate=McpDispatchGate(store,f,"registry-1",peer.registrations("echo",[APPROVAL],f.now))
        def close():
            peer.close()
            backend.close()
        return {"service":create_mcp_proxy_service(gate,f.session["sessionId"],lambda:{"deadline":1800,"cancelled":cancelled}),"close":close}
inner=McpHttpServer(Host(),{**config(endpoint),"callTimeoutMs":3000})
def handle(r):
    if mode=="redirect": return {"status":307,"headers":{"location":"http://127.0.0.1:1/PRIVATE_REDIRECT"},"body":""}
    response=inner.handle(r)
    try: m=json.loads(r["body"])
    except Exception: m={}
    if response["body"] and response["status"]==200 and r["method"]=="POST":
        v=json.loads(response["body"])
        if m.get("method")=="tools/call" and "result" in v:
            if mode=="wrong-id": v["id"]=999
            if mode=="bad-text": v["result"]["content"]=[{"type":"text","text":"PRIVATE_UNCHECKED"}]
            if mode=="oversize": v["result"]["private"]="PRIVATE_"+"x"*1_048_576
            if mode=="session-switch": response["headers"]["mcp-session-id"]="changed"
        response["body"]=json.dumps(v,ensure_ascii=False)
        if mode=="sse":
            response["headers"]["content-type"]="text/event-stream"
            response["body"]=': heartbeat\r\nid: prime\r\ndata:\r\n\r\nevent: message\r\ndata: '+response["body"]+'\r\n\r\n'
        if mode=="sse-notification" and m.get("method")=="tools/call":
            response["headers"]["content-type"]="text/event-stream"
            response["body"]='data: {"jsonrpc":"2.0","method":"notifications/tools/list_changed"}\n\n'+'data: '+response["body"]+'\n\n'
    return response
server.RequestHandlerClass=make_http_handler(SimpleNamespace(handle=handle))
worker=threading.Thread(target=server.serve_forever,daemon=True)
worker.start()
print(endpoint,flush=True)
try: sys.stdin.read()
finally:
    inner.close()
    server.shutdown()
    server.server_close()
    if f: f.close()
